bubbles/main.py

```diff
@@ -0,0 +1,192 @@
+from bubbles.tl_commands import enable_tl_jobs
+from bubbles.tl_utils import tl
+
+PluginManager: PM
+log = logging.getLogger(__name__)
+
+"""
+Notes:
+
+Any long-running response _must_ take in `ack` from the event and call it
+before starting the processing. This is so that Slack doesn't think that
+we didn't hear it and retry the event, which will just screw up processing.
+If we think that something will return immediately (or very close to
+immediately) then we shouldn't need to call `ack()`.
+
+Example:
+@app.event("app_mention")
+def do_long_stuff(ack, say):
+    # note that we're taking in mystery args `say` and `ack`
+    ack()
+    time.sleep(5)
+    say("whew, that took a long time!")
+
+Full list of available helper arguments is here:
+    https://github.com/slackapi/bolt-python#making-things-happen
+
+Full list of available event keys:
+    https://api.slack.com/events
+"""
+
+
+@app.event("reaction_removed")
+@app.event("app_mention")
+def handle(ack):
+    """
+    Gracefully handle extra events so that slack is okay with it.
+
+    Because we listen for direct pings under the `message` event, we don't
+    need to have a handler for `app_mention` events. Unfortunately, if we
+    don't, then slack-bolt spams our logs with "Unhandled request!!!" for
+    `app_mention`. So... we'll just accept `app_mention` events and sinkhole
+    them.
+
+    Same goes for "reaction_removed" -- we need "reaction_added" but don't
+    currently care about removals. If we ever start caring about removals,
+    then we have the handler and can move it somewhere else.
+    """
+    ack()
+
+
+@app.event("message")
+def handle_message(ack, payload, client, context, say, body: dict):
+    # Extract the thread that the message was posted in (if any)
+    thread_ts = body["event"].get("thread_ts")
+
+    def thread_say(*args, **kwargs):
+        """Reply in the thread if the message was sent in a thread."""
+        say(*args, thread_ts=thread_ts, **kwargs)
+
+    # Actually put the slack event handler on this one so that we can dual-wield
+    # `message_received` for interactive mode and normal interaction.
+    PluginManager.message_received(ack, payload, client, context, say=thread_say)
+
+
+@app.event("reaction_added")
+def reaction_added(ack, payload):
+    ack()
+    reaction_added_callback(payload)
+
+
+@click.group(
+    context_settings=dict(help_option_names=["-h", "--help", "--halp"]),
+    invoke_without_command=True,
+)
+@click.pass_context
+@click.option(
+    "-c",
+    "--command",
+    "command",
+    help="Try and run a Bubbles command by name.",
+)
+@click.option(
+    "-i",
+    "--interactive",
+    "interactive",
+    is_flag=True,
+    default=False,
+    help="Start Bubbles in interactive mode for local testing.",
+)
+@click.version_option(version=__version__, prog_name="BubblesV2")
+def main(ctx: Context, command: str, interactive: bool) -> None:
+    """Run Bubbles."""
+    global PluginManager
```

Comment:
This shouldn't be needed because of the declaration at the start (AFAIK)

Reply:
Just re-tested this to make sure: we need the instance of `plugin_manager` created in this function over in `handle_message`, and `handle_message` gets it from the `plugin_manager: PM` line at the top of the file. Having it be global in `main()` makes it so that when `main()` instantiates it, it's now accessible at the top level so `handle_message()` can pick it up.